File: GameServer/GameServerApp/GameServerApp/Proto/CreateProto/GameServerOnePageRequestProto.cs
//===================================================
//作    者：xxx
//创建时间：2019-05-18 15:13:23
//备    注：
//===================================================
using System.Collections;
using System.Collections.Generic;
using System;

/// <summary> 请求服务器单组列表(请求) </summary>
public struct GameServerOnePageRequestProto : IProto
{
    public ushort ProtoCode { get { return 10112; } }

    /// <summary> 页数 </summary>
    public int PageIndex;

    public byte[] ToArray()
    {
        using (MMO_MemoryStream ms = new MMO_MemoryStream())
        {
            ms.WriteUShort(ProtoCode);
            ms.WriteInt(PageIndex);
            return ms.ToArray();
        }
    }

    public static GameServerOnePageRequestProto GetProto(byte[] buffer)
    {
        GameServerOnePageRequestProto proto = new GameServerOnePageRequestProto();
        using (MMO_MemoryStream ms = new MMO_MemoryStream(buffer))
        {
            proto.PageIndex = ms.ReadInt();
        }
        return proto;
    }
}

File: GameServer/GameServerApp/GameServerApp/Proto/CreateLuaProto/GameServerOnePageRequestProto.lua
--请求服务器单组列表(请求)
GameServerOnePageRequestProto = { ProtoCode = 10112, PageIndex = 0 }

--这句是重定义元表的索引，就是说有了这句，这个才是一个类
GameServerOnePageRequestProto.__index = GameServerOnePageRequestProto;

function GameServerOnePageRequestProto.New()
    local self = { }; --初始化self
    setmetatable(self, GameServerOnePageRequestProto); --将self的元表设定为Class
    return self;
end


--发送协议
function GameServerOnePageRequestProto.SendProto(proto)

    local ms = CS.LuaHelper.Instance:CreateMemoryStream();
    ms:WriteUShort(proto.ProtoCode);

    ms:WriteInt(proto.PageIndex);

    CS.LuaHelper.Instance:SendProto(ms:ToArray());
    ms:Dispose();
end


--解析协议
function GameServerOnePageRequestProto.GetProto(buffer)

    local proto = GameServerOnePageRequestProto.New(); --实例化一个协议对象
    local ms = CS.LuaHelper.Instance:CreateMemoryStream(buffer);

    proto.PageIndex = ms:ReadInt();

    ms:Dispose();
    return proto;
end